File: app/Sources/LocalCaptionKit/SpeechStream.swift
import Foundation

public struct CaptionWord: Equatable, Sendable {
    public var text: String
    public var start: Double
    public var end: Double
    public init(_ text: String, start: Double, end: Double) {
        self.text = text; self.start = start; self.end = end
    }
    var midpoint: Double { (start + end) / 2 }
    var normalized: String {
        text.lowercased().trimmingCharacters(in: .whitespacesAndNewlines)
            .trimmingCharacters(in: .punctuationCharacters)
    }
}

public struct SpeechRequest: Sendable {
    public let session: UUID
    public let utterance: Int
    public let startSample: Int
    public let endSample: Int
    public let audio: [Float]
    public let isFinal: Bool
    public let submittedAt: Double
    public init(session: UUID, utterance: Int, startSample: Int, audio: [Float],
                isFinal: Bool, submittedAt: Double = ProcessInfo.processInfo.systemUptime) {
        self.session = session; self.utterance = utterance; self.startSample = startSample
        self.endSample = startSample + audio.count; self.audio = audio
        self.isFinal = isFinal; self.submittedAt = submittedAt
    }
}

public enum SpeechOutcome: Sendable {
    case success(text: String, words: [CaptionWord], fallbacks: Int)
    case empty, filtered, cancelled, timedOut
    case failure(String)

    public var label: String {
        switch self {
        case .success: return "success"
        case .empty: return "empty"
        case .filtered: return "filtered"
        case .cancelled: return "cancelled"
        case .timedOut: return "timeout"
        case .failure: return "failure"
        }
    }
}

/// Pure frame/VAD logic, also used by replay tests. No inference or main-actor work.
/// Retains 200 ms of pre-roll during silence, never a whole silent utterance.
public struct SpeechSegmenter: Sendable {
    public struct Transition: Sendable {
        public let utterance: Int
        public let sample: Int
        public let started: Bool
    }
    public struct Tuning: Sendable {
        public var endpointMs: Int
        public var intervalMs: Int
        public var maxUtteranceS: Int
        public var threshold: Float
        public init(endpointMs: Int = 600, intervalMs: Int = 500,
                    maxUtteranceS: Int = 20, threshold: Float = 0.015) {
            self.endpointMs = max(100, endpointMs)
            self.intervalMs = max(100, intervalMs)
            self.maxUtteranceS = min(60, max(5, maxUtteranceS))
            self.threshold = threshold
        }
    }

    public let session: UUID
    public let tuning: Tuning
    public private(set) var totalSamples: Int
    private var nextUtterance = 0
    private var remainder: [Float] = []
    private var preRoll: [Float] = []
    private var utterance: [Float] = []
    private var startSample = 0
    private var silenceSamples = 0
    private var sinceInterim = 0
    private var transitions: [Transition] = []
    public init(session: UUID, tuning: Tuning = Tuning(), startSample: Int = 0) {
        self.session = session; self.tuning = tuning; self.totalSamples = startSample
    }

    public mutating func append(_ samples: [Float], now: Double) -> [SpeechRequest] {
        remainder.append(contentsOf: samples)
        var requests: [SpeechRequest] = []
        var offset = 0
        while remainder.count - offset >= 1600 {
            consume(Array(remainder[offset..<(offset + 1600)]), now: now, into: &requests)
            offset += 1600
        }
        remainder = Array(remainder.dropFirst(offset))
        return requests
    }

    /// Capture must stop before this is called. Includes the last sub-100 ms frame.
    public mutating func finish(now: Double) -> [SpeechRequest] {
        var requests: [SpeechRequest] = []
        if !remainder.isEmpty {
            let tail = remainder; remainder = []
            consume(tail, now: now, into: &requests)
        }
        if !utterance.isEmpty { finalize(now: now, into: &requests) }
        preRoll = []
        return requests
    }

    public mutating func drainTransitions() -> [Transition] {
        let events = transitions; transitions = []
        return events
    }

    private mutating func consume(_ frame: [Float], now: Double, into requests: inout [SpeechRequest]) {
        let speaking = Self.rms(frame) >= tuning.threshold
        let frameStart = totalSamples
        totalSamples += frame.count
        if utterance.isEmpty {
            guard speaking else {
                preRoll = Array((preRoll + frame).suffix(3200))
                return
            }
            startSample = frameStart - preRoll.count
            utterance = preRoll; preRoll = []
            transitions.append(Transition(utterance: nextUtterance, sample: frameStart, started: true))
        }
        utterance.append(contentsOf: frame)
        silenceSamples = speaking ? 0 : silenceSamples + frame.count
        sinceInterim += frame.count
        if speaking && sinceInterim >= tuning.intervalMs * 16 {
            sinceInterim = 0
            let tail = Array(utterance.suffix(96000))
            requests.append(SpeechRequest(session: session, utterance: nextUtterance,
                startSample: totalSamples - tail.count, audio: tail, isFinal: false, submittedAt: now))
        }
        if silenceSamples >= tuning.endpointMs * 16 || utterance.count >= tuning.maxUtteranceS * 16000 {
            finalize(now: now, into: &requests)
        }
    }

    private mutating func finalize(now: Double, into requests: inout [SpeechRequest]) {
        transitions.append(Transition(utterance: nextUtterance, sample: totalSamples, started: false))
        requests.append(SpeechRequest(session: session, utterance: nextUtterance,
            startSample: startSample, audio: utterance, isFinal: true, submittedAt: now))
        nextUtterance += 1; utterance = []; silenceSamples = 0; sinceInterim = 0
    }

    public static func rms(_ samples: [Float]) -> Float {
        guard !samples.isEmpty else { return 0 }
        return (samples.reduce(Float(0)) { $0 + $1 * $1 } / Float(samples.count)).squareRoot()
    }
}

/// Merge model word timings in session coordinates, never by the length of an old
/// prefix. The overlapping suffix is provisional and may be corrected by a new decode.
public struct RollingCaption: Sendable {
    public private(set) var words: [CaptionWord] = []
    public private(set) var lastEndSample = -1
    private var lastStartSample = -1
    public init() {}
    public var text: String { words.map(\.text).joined(separator: " ") }

    @discardableResult
    public mutating func update(_ incoming: [CaptionWord], startSample: Int, endSample: Int) -> Bool {
        guard endSample > lastEndSample, !incoming.isEmpty else { return false }
        let offset = Double(startSample) / 16000
        let duration = Double(endSample - startSample) / 16000
        let valid = incoming.filter {
            !$0.text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty &&
            $0.start.isFinite && $0.end.isFinite && $0.start >= 0 &&
            $0.end >= $0.start && $0.start <= duration
        }.map { CaptionWord($0.text.trimmingCharacters(in: .whitespacesAndNewlines),
                            start: offset + $0.start, end: offset + min(duration, $0.end)) }
        guard !valid.isEmpty else { return false }

        if words.isEmpty || startSample == lastStartSample {
            // Same audio origin: permit corrections, including shorter hypotheses.
            let prefix = words.prefix { $0.end <= offset }
            words = Array(prefix) + valid
        } else {
            // Anchor on the first matching word within 350 ms. Time proximity
            // distinguishes real repeated phrases from duplicated overlap.
            var anchor: (old: Int, new: Int)?
            for (j, new) in valid.enumerated() {
                let candidates = words.indices.filter {
                    words[$0].normalized == new.normalized && !new.normalized.isEmpty &&
                    abs(words[$0].midpoint - new.midpoint) <= 0.35
                }
                if let i = candidates.min(by: {
                    abs(words[$0].midpoint - new.midpoint) < abs(words[$1].midpoint - new.midpoint)
                }) { anchor = (i, j); break }
            }
            if let anchor {
                // Keep the newer model's prefix where it contains words before the
                // anchor, preserving only the old words outside that covered range.
                let prefixCount = anchor.new == 0 ? anchor.old :
                    min(anchor.old, words.prefix { $0.midpoint < valid[0].start }.count)
                words = Array(words.prefix(prefixCount)) + valid
            } else {
                // No lexical anchor: the new window owns its time range. This is an
                // explicit provisional correction, not an invented textual overlap.
                words = Array(words.prefix { $0.midpoint < valid[0].start }) + valid
            }
        }
        lastStartSample = startSample; lastEndSample = endSample
        return true
    }
}

File: windows/src/LocalCaption.Core/Audio/SpeechStream.cs
using System.Diagnostics;
using System.Globalization;

namespace LocalCaption.Core.Audio;

/// <summary>
/// Monotonic seconds, the C# equivalent of <c>ProcessInfo.processInfo.systemUptime</c>.
/// Wall-clock time must never be used for scheduling: the ASUS resyncs its clock over the
/// network and a backward jump would make decode budgets and backlog maths nonsense.
/// </summary>
public static class MonotonicClock
{
    public static double Now => (double)Stopwatch.GetTimestamp() / Stopwatch.Frequency;
}

/// <summary>One word with its timing in session coordinates (seconds).</summary>
public readonly record struct CaptionWord(string Text, double Start, double End)
{
    public double Midpoint => (Start + End) / 2;

    /// <summary>
    /// Lowercase, whitespace-trimmed, then punctuation-trimmed — in that order, so
    /// <c>"hello, "</c> and <c>"Hello"</c> compare equal. Used only to anchor overlapping
    /// hypotheses against each other, never for display.
    /// </summary>
    public string Normalized
    {
        get
        {
            var t = Text.ToLowerInvariant().Trim();
            var start = 0;
            var end = t.Length;
            while (start < end && char.IsPunctuation(t[start])) start++;
            while (end > start && char.IsPunctuation(t[end - 1])) end--;
            return t[start..end];
        }
    }
}

/// <summary>One decode request: a window of audio plus where it sits in the session.</summary>
public sealed class SpeechRequest
{
    public Guid Session { get; }
    public int Utterance { get; }
    public int StartSample { get; }
    public int EndSample { get; }
    public IReadOnlyList<float> Audio { get; }
    public bool IsFinal { get; }
    public double SubmittedAt { get; }

    public SpeechRequest(Guid session, int utterance, int startSample, IReadOnlyList<float> audio,
                         bool isFinal, double? submittedAt = null)
    {
        Session = session;
        Utterance = utterance;
        StartSample = startSample;
        EndSample = startSample + audio.Count;
        Audio = audio;
        IsFinal = isFinal;
        SubmittedAt = submittedAt ?? MonotonicClock.Now;
    }
}

/// <summary>What a decode produced. <see cref="Label"/> is what lands in the metrics log.</summary>
public abstract record SpeechOutcome
{
    private SpeechOutcome() { }

    public sealed record Success(string Text, IReadOnlyList<CaptionWord> Words, int Fallbacks) : SpeechOutcome;
    public sealed record Empty : SpeechOutcome;
    public sealed record Filtered : SpeechOutcome;
    public sealed record Cancelled : SpeechOutcome;
    public sealed record TimedOut : SpeechOutcome;
    public sealed record Failure(string Message) : SpeechOutcome;

    public string Label => this switch
    {
        Success => "success",
        Empty => "empty",
        Filtered => "filtered",
        Cancelled => "cancelled",
        TimedOut => "timeout",
        Failure => "failure",
        _ => "unknown",
    };
}

/// <summary>
/// Pure frame/VAD logic — no inference, no UI, no audio device. Retains 200 ms of pre-roll
/// during silence so an utterance is never clipped at its onset, but never buffers a whole
/// silent stretch: five minutes of silence must enqueue no work at all.
/// </summary>
public sealed class SpeechSegmenter
{
    public readonly record struct Transition(int Utterance, int Sample, bool Started);

    public readonly record struct Tuning
    {
        public int EndpointMs { get; }
        public int IntervalMs { get; }
        public int MaxUtteranceS { get; }
        public float Threshold { get; }

        public Tuning(int endpointMs = 600, int intervalMs = 500,
                      int maxUtteranceS = 20, float threshold = 0.015f)
        {
            EndpointMs = Math.Max(100, endpointMs);
            IntervalMs = Math.Max(100, intervalMs);
            MaxUtteranceS = Math.Min(60, Math.Max(5, maxUtteranceS));
            Threshold = threshold;
        }
    }

    private const int SampleRate = 16000;
    private const int FrameSamples = 1600;      // 100 ms
    private const int PreRollSamples = 3200;    // 200 ms
    private const int InterimTailSamples = 96000; // 6 s

    public Guid Session { get; }
    public Tuning Settings { get; }
    public int TotalSamples { get; private set; }

    private int _nextUtterance;
    private readonly List<float> _remainder = [];
    private readonly List<float> _preRoll = [];
    private readonly List<float> _utterance = [];
    private int _startSample;
    private int _silenceSamples;
    private int _sinceInterim;
    private readonly List<Transition> _transitions = [];

    public SpeechSegmenter(Guid session, Tuning? tuning = null, int startSample = 0)
    {
        Session = session;
        Settings = tuning ?? new Tuning();
        TotalSamples = startSample;
    }

    public IReadOnlyList<SpeechRequest> Append(IReadOnlyList<float> samples, double now)
    {
        _remainder.AddRange(samples);
        var requests = new List<SpeechRequest>();
        var offset = 0;
        while (_remainder.Count - offset >= FrameSamples)
        {
            Consume(_remainder.GetRange(offset, FrameSamples), now, requests);
            offset += FrameSamples;
        }
        _remainder.RemoveRange(0, offset);
        return requests;
    }

    /// <summary>
    /// Capture must be stopped before this is called. Includes the last sub-100 ms frame, so
    /// no speech sample is ever dropped at Stop.
    /// </summary>
    public IReadOnlyList<SpeechRequest> Finish(double now)
    {
        var requests = new List<SpeechRequest>();
        if (_remainder.Count > 0)
        {
            var tail = new List<float>(_remainder);
            _remainder.Clear();
            Consume(tail, now, requests);
        }
        if (_utterance.Count > 0) Finalize(now, requests);
        _preRoll.Clear();
        return requests;
    }

    public IReadOnlyList<Transition> DrainTransitions()
    {
        var events = _transitions.ToArray();
        _transitions.Clear();
        return events;
    }

    private void Consume(List<float> frame, double now, List<SpeechRequest> requests)
    {
        var speaking = Rms(frame) >= Settings.Threshold;
        var frameStart = TotalSamples;
        TotalSamples += frame.Count;

        if (_utterance.Count == 0)
        {
            if (!speaking)
            {
                _preRoll.AddRange(frame);
                if (_preRoll.Count > PreRollSamples) _preRoll.RemoveRange(0, _preRoll.Count - PreRollSamples);
                return;
            }
            _startSample = frameStart - _preRoll.Count;
            _utterance.AddRange(_preRoll);
            _preRoll.Clear();
            _transitions.Add(new Transition(_nextUtterance, frameStart, Started: true));
        }

        _utterance.AddRange(frame);
        _silenceSamples = speaking ? 0 : _silenceSamples + frame.Count;
        _sinceInterim += frame.Count;

        if (speaking && _sinceInterim >= Settings.IntervalMs * 16)
        {
            _sinceInterim = 0;
            var take = Math.Min(InterimTailSamples, _utterance.Count);
            var tail = _utterance.GetRange(_utterance.Count - take, take);
            requests.Add(new SpeechRequest(Session, _nextUtterance, TotalSamples - tail.Count,
                                           tail, isFinal: false, submittedAt: now));
        }

        if (_silenceSamples >= Settings.EndpointMs * 16 ||
            _utterance.Count >= Settings.MaxUtteranceS * SampleRate)
        {
            Finalize(now, requests);
        }
    }

    private void Finalize(double now, List<SpeechRequest> requests)
    {
        _transitions.Add(new Transition(_nextUtterance, TotalSamples, Started: false));
        requests.Add(new SpeechRequest(Session, _nextUtterance, _startSample,
                                       _utterance.ToArray(), isFinal: true, submittedAt: now));
        _nextUtterance++;
        _utterance.Clear();
        _silenceSamples = 0;
        _sinceInterim = 0;
    }

    public static float Rms(IReadOnlyList<float> samples)
    {
        if (samples.Count == 0) return 0;
        float sum = 0;
        for (var i = 0; i < samples.Count; i++) sum += samples[i] * samples[i];
        return MathF.Sqrt(sum / samples.Count);
    }
}
